### PR TITLE
Ignore Links for Rules

### LINKED ISSUE #196: [Usability-breaking issue] Linter modifies links
I have files with multiple spaces in their names, because reasons. Using Linter breaks the links to all those files by removing the double spaces. I'm sure this applies to other filters, and is something that completely breaks the usability of the plugin.

And yes, I can remove the double spaces from the file names, but Linter absolutely shouldn't be messing with anything that's inside a `link` environment anyways.

### PR DESCRIPTION
**Pull request by pjkaufman**
@platers 

Fixes #196

I went ahead and added a the regex for links and the UTs to make sure the 3 kinds of links lint correctly. However a couple of situations did not work for me, but I am not sure if that was intended:
- `[  ]` does not have the extra spaces removed from it for some reason
- `(  )` also does not have the extra spaces removed from it for some reason

I added a links placeholder which had to have English letters so the English and Chinese spacing lint rule would still work.
I renamed `ignoreCodeBlocksAndYAML` to `ignoreCodeBlocksYAMLAndLinks` since it now does more things.

Please let me know if we need to not remove the empty spaces from the two above scenarios that I hit issues with or if there is a different way we would like to ignore the links before any rules are added around them (#32  is one that would require reworking this so that the rules for links would work just fine while the rules for other operations would not affect it).

I am not sure if remark has a link position option like for the code blocks, but if it does, please let me know and I can try to use that if we would like.

### CODE REVIEW

File: src/test.ts [hunk 1]
```diff
@@ -674,3 +674,51 @@ describe('YAML Title', () => {
...
+    ![This has  spaces in it](File with  spaces.png)
+    `;
+
+    expect(rulesDict['trailing-spaces'].apply(before)).toBe(after);
+  });
+  it('Wiki link with spaces stays the same', () => {
+    const before = dedent`
+    # Hello world
+
+    [This has  spaces in it](File with  spaces.md)
```
Comment (platers) [inline src/test.ts:713]: should this test `[[wiki    link]]`?

File: src/utils.ts [hunk 2]
```diff
@@ -96,19 +96,23 @@ export function moveFootnotesToEnd(text: string) {
...
  * @return {string} The processed text
  */
-export function ignoreCodeBlocksAndYAML(text: string, func: (text: string) => string): string {
+export function ignoreCodeBlocksYAMLAndLinks(text: string, func: (text: string) => string): string {
   const codePlaceholder = 'PLACEHOLDER 321417';
   const ret = replaceCodeblocks(text, codePlaceholder);
   text = ret.text;
   const replacedCodeBlocks = ret.replacedCodeBlocks;
 
+  const linkPlaceHolder = 'PLACEHOLDER_LINK 57849';
```
Comment (platers) [inline src/utils.ts:112]: I think we should replace YAML before links, if there is a link in the YAML we dont want to touch it at all.

**Review by platers — COMMENTED** [review]
thanks!

File: src/test.ts (hunk 1 above)
Comment (pjkaufman) [reply src/test.ts]: That is correct. I have updated the test.

File: src/utils.ts (hunk 2 above)
Comment (pjkaufman) [reply src/utils.ts]: Sounds good. I have swapped the order.

**Review by platers — APPROVED** [review]

**Pull request merged**